packages/components/src/slot-fill/provider.tsx

```diff
@@ -0,0 +1,129 @@
+ */
+import SlotFillContext from './context';
+import type {
+	FillComponentProps,
+	BaseSlotFillContext,
+	BaseSlotComponentProps,
+	SlotFillProviderProps,
+	SlotKey,
+} from './types';
+
+function createSlotRegistry(): BaseSlotFillContext {
+	const slots: Record< SlotKey, Component< BaseSlotComponentProps > > = {};
+	const fills: Record< SlotKey, FillComponentProps[] > = {};
+	let listeners: Array< () => void > = [];
+
+	function registerSlot(
+		name: SlotKey,
+		slot: Component< BaseSlotComponentProps >
+	) {
+		const previousSlot = slots[ name ];
+		slots[ name ] = slot;
+		triggerListeners();
+
+		// Sometimes the fills are registered after the initial render of slot
+		// But before the registerSlot call, we need to rerender the slot.
+		forceUpdateSlot( name );
+
+		// If a new instance of a slot is being mounted while another with the
+		// same name exists, force its update _after_ the new slot has been
+		// assigned into the instance, such that its own rendering of children
+		// will be empty (the new Slot will subsume all fills for this name).
+		if ( previousSlot ) {
+			previousSlot.forceUpdate();
+		}
+	}
+
+	function registerFill( name: SlotKey, instance: FillComponentProps ) {
+		fills[ name ] = [ ...( fills[ name ] || [] ), instance ];
+		forceUpdateSlot( name );
+	}
+
+	function unregisterSlot(
+		name: SlotKey,
+		instance: Component< BaseSlotComponentProps >
+	) {
+		// If a previous instance of a Slot by this name unmounts, do nothing,
+		// as the slot and its fills should only be removed for the current
+		// known instance.
+		if ( slots[ name ] !== instance ) {
+			return;
+		}
+
+		delete slots[ name ];
+		triggerListeners();
+	}
+
+	function unregisterFill( name: SlotKey, instance: FillComponentProps ) {
+		fills[ name ] =
+			fills[ name ]?.filter( ( fill ) => fill !== instance ) ?? [];
+		forceUpdateSlot( name );
+	}
+
+	function getSlot(
+		name: SlotKey
+	): Component< BaseSlotComponentProps > | undefined {
+		return slots[ name ];
+	}
+
+	function getFills(
+		name: SlotKey,
+		slotInstance: Component< BaseSlotComponentProps >
+	): FillComponentProps[] {
+		// Fills should only be returned for the current instance of the slot
+		// in which they occupy.
+		if ( slots[ name ] !== slotInstance ) {
+			return [];
+		}
+		return fills[ name ];
+	}
+
+	function forceUpdateSlot( name: SlotKey ) {
+		const slot = getSlot( name );
+
+		if ( slot ) {
+			slot.forceUpdate();
+		}
+	}
+
+	function triggerListeners() {
+		listeners.forEach( ( listener ) => listener() );
+	}
+
+	function subscribe( listener: () => void ) {
+		listeners.push( listener );
+
+		return () => {
+			listeners = listeners.filter( ( l ) => l !== listener );
+		};
+	}
+
+	return {
+		registerSlot,
+		unregisterSlot,
+		registerFill,
+		unregisterFill,
+		getSlot,
+		getFills,
+		subscribe,
+	};
+}
+
+export function SlotFillProvider( { children }: SlotFillProviderProps ) {
+	const [ contextValue ] = useState( createSlotRegistry );
```

Comment:
This is an interesting way to use `useState`, I've actually never seen this pattern before! But I feel like doing a `useMemo` is more idiomatic and captures our intent better. What do you think? Are you aware of any reasons for preferring a `useState` here?

```suggestion
	const contextValue = useMemo( createSlotRegistry, [] );
```

Reply:
https://github.com/WordPress/gutenberg/pull/51350/files/ef5a64d427953f06b7c2a5e92ea92415a647c9d1#diff-e0f964e993ca91fda60b836c0e5e6286b5af387e8f4d632d20a366b28a8da133L89

This is the same implementation as this line. I too think it would be easier to understand using `useMemo`.

Reply:
Agreed with both, I think we can switch to a `useMemo`  here

Reply:
Switched to useMemo.